<?xml version='1.0' encoding='UTF-8'?>
<root>
 <tabbed_widget parent="main_window" name="Main Window">
  <Tab tab_name="Reference vs Actual" containers="1">
   <Container>
    <DockSplitter sizes="0.500248;0.499752" orientation="-" count="2">
     <DockArea name="Trailing Error">
      <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
       <range left="201.897373" right="206.908213" bottom="0.899641" top="5.114720"/>
       <limitY/>
       <curve color="#1f77b4" name="/trailing/gap_data/actual"/>
       <curve color="#d62728" name="/trailing/gap_data/should"/>
      </plot>
     </DockArea>
     <DockArea name="Velocity Error">
      <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
       <range left="201.897373" right="206.908213" bottom="2.084750" top="5.935211"/>
       <limitY/>
       <curve color="#ff7f0e" name="/velocity_pid/should"/>
       <curve color="#1ac938" name="/velocity_pid/actual"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="Trailing" containers="1">
   <Container>
    <DockSplitter sizes="0.500248;0.499752" orientation="-" count="2">
     <DockArea name="...">
      <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
       <range left="201.897373" right="206.896486" bottom="0.899641" top="5.114720"/>
       <limitY/>
       <curve color="#1f77b4" name="/trailing/gap_data/actual"/>
       <curve color="#d62728" name="/trailing/gap_data/should"/>
      </plot>
     </DockArea>
     <DockSplitter sizes="0.5;0.5" orientation="|" count="2">
      <DockArea name="...">
       <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
        <range left="201.897373" right="206.896486" bottom="-0.100000" top="0.100000"/>
        <limitY/>
        <curve color="#f14cc1" name="/trailing/gap_data/i_value"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
        <range left="201.897373" right="206.896486" bottom="2.126777" top="5.828964"/>
        <limitY/>
        <curve color="#9467bd" name="/trailing/gap_data/d_value"/>
       </plot>
      </DockArea>
     </DockSplitter>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="Velocity" containers="1">
   <Container>
    <DockSplitter sizes="0.500248;0.499752" orientation="-" count="2">
     <DockArea name="...">
      <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
       <range left="201.933027" right="206.908213" bottom="2.084750" top="5.935211"/>
       <limitY/>
       <curve color="#1ac938" name="/velocity_pid/actual"/>
       <curve color="#0e6bff" name="/velocity_pid/should"/>
      </plot>
     </DockArea>
     <DockSplitter sizes="0.5;0.5" orientation="|" count="2">
      <DockArea name="...">
       <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
        <range left="201.933027" right="206.908213" bottom="-1.050000" top="1.050000"/>
        <limitY/>
        <curve color="#17becf" name="/velocity_pid/i_value"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot flip_x="false" style="Lines" flip_y="false" mode="TimeSeries">
        <range left="201.933027" right="206.908213" bottom="-26.549971" top="11.852523"/>
        <limitY/>
        <curve color="#bcbd22" name="/velocity_pid/d_value"/>
       </plot>
      </DockArea>
     </DockSplitter>
    </DockSplitter>
   </Container>
  </Tab>
  <currentTabIndex index="1"/>
 </tabbed_widget>
 <use_relative_time_offset enabled="1"/>
 <!-- - - - - - - - - - - - - - - -->
 <!-- - - - - - - - - - - - - - - -->
 <Plugins>
  <plugin ID="DataLoad CSV">
   <default delimiter="0" time_axis=""/>
  </plugin>
  <plugin ID="DataLoad MCAP"/>
  <plugin ID="DataLoad ROS bags">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
  </plugin>
  <plugin ID="DataLoad ULog"/>
  <plugin ID="ROS Topic Subscriber">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="120"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
  </plugin>
  <plugin ID="UDP Server"/>
  <plugin ID="WebSocket Server"/>
  <plugin ID="ZMQ Subscriber"/>
  <plugin ID="Fast Fourier Transform"/>
  <plugin ID="Quaternion to RPY"/>
  <plugin ID="Reactive Script Editor">
   <library code="--[[ Helper function to create a series from arrays&#xa;&#xa; new_series: a series previously created with ScatterXY.new(name)&#xa; prefix:     prefix of the timeseries, before the index of the array&#xa; suffix_X:   suffix to complete the name of the series containing the X value. If [nil], use the index of the array.&#xa; suffix_Y:   suffix to complete the name of the series containing the Y value&#xa; timestamp:   usually the tracker_time variable&#xa;              &#xa; Example:&#xa; &#xa; Assuming we have multiple series in the form:&#xa; &#xa;   /trajectory/node.{X}/position/x&#xa;   /trajectory/node.{X}/position/y&#xa;   &#xa; where {N} is the index of the array (integer). We can create a reactive series from the array with:&#xa; &#xa;   new_series = ScatterXY.new(&quot;my_trajectory&quot;) &#xa;   CreateSeriesFromArray( new_series, &quot;/trajectory/node&quot;, &quot;position/x&quot;, &quot;position/y&quot;, tracker_time );&#xa;--]]&#xa;&#xa;function CreateSeriesFromArray( new_series, prefix, suffix_X, suffix_Y, timestamp )&#xa;  &#xa;  --- clear previous values&#xa;  new_series:clear()&#xa;  &#xa;  --- Append points to new_series&#xa;  index = 0&#xa;  while(true) do&#xa;&#xa;    x = index;&#xa;    -- if not nil, get the X coordinate from a series&#xa;    if suffix_X ~= nil then &#xa;      series_x = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_X) )&#xa;      if series_x == nil then break end&#xa;      x = series_x:atTime(timestamp)&#x9; &#xa;    end&#xa;    &#xa;    series_y = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_Y) )&#xa;    if series_y == nil then break end &#xa;    y = series_y:atTime(timestamp)&#xa;    &#xa;    new_series:push_back(x,y)&#xa;    index = index+1&#xa;  end&#xa;end&#xa;&#xa;--[[ Similar to the built-in function GetSeriesNames(), but select only the names with a give prefix. --]]&#xa;&#xa;function GetSeriesNamesByPrefix(prefix)&#xa;  -- GetSeriesNames(9 is a built-in function&#xa;  all_names = GetSeriesNames()&#xa;  filtered_names = {}&#xa;  for i, name in ipairs(all_names)  do&#xa;    -- check the prefix&#xa;    if name:find(prefix, 1, #prefix) then&#xa;      table.insert(filtered_names, name);&#xa;    end&#xa;  end&#xa;  return filtered_names&#xa;end&#xa;&#xa;--[[ Modify an existing series, applying offsets to all their X and Y values&#xa;&#xa; series: an existing timeseries, obtained with TimeseriesView.find(name)&#xa; delta_x: offset to apply to each x value&#xa; delta_y: offset to apply to each y value &#xa;  &#xa;--]]&#xa;&#xa;function ApplyOffsetInPlace(series, delta_x, delta_y)&#xa;  -- use C++ indeces, not Lua indeces&#xa;  for index=0, series:size()-1 do&#xa;    x,y = series:at(index)&#xa;    series:set(index, x + delta_x, y + delta_y)&#xa;  end&#xa;end&#xa;"/>
   <scripts/>
  </plugin>
  <plugin ID="CSV Exporter"/>
  <plugin ID="ROS /rosout Visualization"/>
  <plugin ID="ROS Topic Re-Publisher"/>
 </Plugins>
 <!-- - - - - - - - - - - - - - - -->
 <previouslyLoaded_Datafiles/>
 <previouslyLoaded_Streamer name="ROS Topic Subscriber"/>
 <!-- - - - - - - - - - - - - - - -->
 <customMathEquations/>
 <snippets/>
 <!-- - - - - - - - - - - - - - - -->
</root>

